### PR TITLE
fix(cli): fail incomplete uninstall scans

### LINKED ISSUE #767: language uninstall --all silently skips unreadable install directories
## Summary

`kakehashi language uninstall --all` ignores `read_dir` failures and flattens per-entry failures while discovering installed parsers and queries.

## User impact

An unreadable install directory can make `--all` omit installed languages and still report success (including `No languages installed to uninstall`), leaving files behind despite the command contract.

## Expected behavior

Missing install directories may be treated as empty, but other directory-open or entry-iteration errors should fail the operation before claiming completion.

## Scope

Tracked separately from the same scan bug in `language status` so each PR can land independently.

### PR DESCRIPTION
**Pull request by atusy**
## Summary

- treat only genuinely missing parser/query roots as an empty install during `language uninstall`
- propagate directory-open, per-entry, metadata, symlink, and query-recovery errors before removal
- preflight both roots before bulk recovery; scope targeted recovery/preflight to the requested language
- recover only complete, owned backups and preserve idempotent behavior when roots vanish concurrently

## User impact

`language uninstall` no longer reports success while unreadable or malformed install state remains behind, and scan/recovery failures cannot partially delete already-discovered parsers or unrelated language state.

## Validation

- RED regressions cover unreadable and dangling roots, malformed entries, symlinked roots/recovery entries, recovery failure ordering, targeted recovery scope, and unreadable backup metadata
- `cargo test --test test_cli` (61 passed)
- `cargo test --lib install::queries::tests` (27 passed)
- `cargo clippy --bin kakehashi --lib -- -D warnings`
- `cargo fmt --check`
- fresh exact-head Codex review: no actionable findings

## Follow-up

- #774 tracks global coordination when installs race `uninstall --all`; it requires an operation-wide protocol beyond this static filesystem scan fix.

Closes #767

## Summary by CodeRabbit

* **Bug Fixes**
  * Hardened language uninstall with stronger preflight checks of parser/query paths, including handling unreadable, dangling, and unsafe symlink scenarios.
  * Improved recovery of interrupted query installs, including targeted recovery for a single language and safer `--all` recovery/rescanning so recovered items are included.
  * Updated removal sequencing to attempt query/backup removal before parser deletion, preserving parser data if query removal fails.
  * Added stricter validation for recovery/backup directory naming and completeness.
* **Tests**
  * Expanded Unix integration tests covering uninstall `--all`/targeted edge cases, permission failures, symlinks/dangling entries, and recovery sequencing rules.

### CODE REVIEW

File: tests/test_cli.rs [hunk 1]
```diff
@@ -1209,6 +1209,177 @@ fn test_language_uninstall_all() {
...
+            "--force",
+        ])
+        .output()
+        .expect("Failed to execute command");
+
+    assert!(
+        !output.status.success(),
+        "a present but unreadable install root must fail; stderr: {}",
+        String::from_utf8_lossy(&output.stderr)
+    );
```
[the remark this answers is not in the document]
Comment (atusy) [reply tests/test_cli.rs:1317]: Fixed in f5eaf384f: the assertion now names the dangling install root, matching the symlink fixture.

File: src/bin/main.rs [hunk 2]
```diff
@@ -480,13 +484,98 @@ fn installed_query_language_name(path: &Path) -> Option<String> {
...
+fn installed_query_language_name_for_uninstall(
+    entry: &std::fs::DirEntry,
+    queries_dir: &Path,
+) -> Result<Option<String>, String> {
+    let file_type = entry.file_type().map_err(|e| {
+        format!(
+            "cannot inspect an entry in queries directory '{}': {e}",
+            queries_dir.display()
+        )
+    })?;
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/bin/main.rs]: Fixed in 90422425a: query entry classification captures the concrete entry path once and includes it in file-type and symlink-resolution failures.

File: tests/test_cli.rs [hunk 3]
```diff
@@ -1209,6 +1209,221 @@ fn test_language_uninstall_all() {
...
+            "--data-dir",
+            test_dir.path().to_str().unwrap(),
+            "--force",
+        ])
+        .output()
+        .expect("Failed to execute command");
+
+    assert!(
+        !output.status.success(),
+        "an unreadable query entry must fail; stderr: {}",
```
[the remark this answers is not in the document]
Comment (atusy) [reply tests/test_cli.rs]: Superseded in acddede8c: the symlink loop is now correctly treated as a removable query namespace entry, and the rewritten assertion says `query symlink should be removable without resolving its target`.

File: src/bin/main.rs [hunk 4]
```diff
@@ -480,13 +484,133 @@ fn installed_query_language_name(path: &Path) -> Option<String> {
...
+    // make the kakehashi-owned namespace impossible to clean.
+    let is_dir = file_type.is_dir() || file_type.is_symlink();
+    let language = installed_query_language_name_if_dir(&path, is_dir);
+    // Hidden staging/backup directories are skipped as installed languages,
+    // but recovery can mutate them into canonical installs. Validate every
+    // real directory before recovery so an unreadable candidate cannot be
+    // renamed first and fail only on the post-recovery scan.
+    if file_type.is_dir() {
+        preflight_query_install_tree(&path)?;
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/bin/main.rs:570]: Fixed in 12d12e5ee: recognized hidden recovery entries are rejected when symlinks in both bulk and targeted preflight, with a regression proving an outside backup target is untouched.

File: src/bin/main.rs [hunk 5]
```diff
@@ -480,13 +484,133 @@ fn installed_query_language_name(path: &Path) -> Option<String> {
...
+            entry
+                .file_type()
+                .map_err(|e| format!("cannot inspect parser entry '{}': {e}", path.display()))?;
+            let is_parser = path
+                .extension()
+                .map(|ext| ext == std::env::consts::DLL_EXTENSION)
+                .unwrap_or(false);
+            if is_parser && let Some(stem) = path.file_stem() {
+                languages.insert(stem.to_string_lossy().to_string());
+            }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/bin/main.rs:649]: Fixed in 5e966a2cb: bulk discovery now requires parser candidates to be files or symlinks, targeted lookup applies the same rule before query mutation, and directory/socket regressions cover both modes.

File: src/bin/main.rs [hunk 6]
```diff
@@ -497,37 +621,43 @@ fn run_language_uninstall(
...
+        })?;
+    } else {
+        // Targeted uninstall does not need to enumerate every entry, but it
+        // still builds removal paths beneath both roots. Reject a symlinked or
+        // non-directory root before recovery/removal can escape data_dir.
+        for (path, kind) in [(&parser_dir, "parser"), (&queries_dir, "queries")] {
+            read_optional_install_dir(path, kind).map_err(|e| {
+                eprintln!("Failed to validate install roots: {e}");
+                ExitCode::FAILURE
+            })?;
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/bin/main.rs]: Kept the readability check intentionally. Targeted uninstall invokes query recovery, which enumerates the queries root and can mutate staging/backups for multiple languages; metadata-only validation would bypass exactly the unreadable-state failures #767 requires us to surface before recovery. The fixed removal path is therefore not sufficient to make unreadable roots safe.

File: src/install/queries.rs [hunk 7]
```diff
@@ -656,6 +656,14 @@ fn generated_temp_parts(name: &str) -> Option<(&str, &str, &str)> {
     }
 }
 
+/// Whether a directory name belongs to the staging/backup namespace that
+/// [`recover_interrupted_query_installs`] may mutate.
+pub fn is_recovery_directory_name(name: &str) -> bool {
+    generated_backup_parts(name).is_some_and(|(language, _, _)| is_safe_language_name(language))
+        || generated_temp_parts(name)
+            .is_some_and(|(language, _, _)| is_safe_language_name(language))
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/queries.rs:717]: Fixed in f1cf12ecc: recovery now propagates `ReadDir` entry and `file_type` errors and processes only real directories. A direct unit test proves backup symlinks are not followed.

File: src/bin/main.rs [hunk 8]
```diff
@@ -480,13 +484,209 @@ fn installed_query_language_name(path: &Path) -> Option<String> {
...
+                "{kind} path '{}' is not a directory",
+                path.display()
+            ));
+        }
+        Ok(_) => {}
+    }
+    std::fs::read_dir(path)
+        .map(Some)
+        .map_err(|e| format!("cannot read {kind} directory '{}': {e}", path.display()))
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/bin/main.rs]: Fixed in 072ae595f: read_dir NotFound now preserves the missing-root-is-empty contract across the metadata/open race.

File: tests/test_cli.rs [hunk 9]
```diff
@@ -1209,6 +1209,820 @@ fn test_language_uninstall_all() {
...
+    fs::set_permissions(&install_dir, fs::Permissions::from_mode(0o000))
+        .expect("Failed to make install dir unreadable");
+
+    // Elevated test environments may retain permission to read mode-000 paths.
+    // Skip there rather than asserting a failure the OS cannot produce.
+    if fs::read_dir(&install_dir).is_ok() {
+        fs::set_permissions(&install_dir, fs::Permissions::from_mode(0o700))
+            .expect("Failed to restore install dir permissions");
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply tests/test_cli.rs]: Fixed in 954769afc: permission-based tests now consume every read_dir entry before deciding the directory is enumerable.

File: tests/test_cli.rs [hunk 10]
```diff
@@ -1209,6 +1209,820 @@ fn test_language_uninstall_all() {
...
+    let stranded_tmp = test_dir.path().join("queries/.lua.4294967295.0.tmp");
+    fs::create_dir_all(&parser_dir).expect("Failed to create parser dir");
+    fs::create_dir_all(&stranded_tmp).expect("Failed to create stranded query temp dir");
+    fs::set_permissions(&parser_dir, fs::Permissions::from_mode(0o000))
+        .expect("Failed to make parser dir unreadable");
+    if fs::read_dir(&parser_dir).is_ok() {
+        fs::set_permissions(&parser_dir, fs::Permissions::from_mode(0o700))
+            .expect("Failed to restore parser dir permissions");
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply tests/test_cli.rs]: Fixed in 954769afc: permission-based tests now consume every read_dir entry before deciding the directory is enumerable.

File: tests/test_cli.rs [hunk 11]
```diff
@@ -1209,6 +1209,820 @@ fn test_language_uninstall_all() {
...
+    fs::write(&parser, "fake").expect("Failed to write parser");
+    fs::write(query_dir.join("highlights.scm"), "(comment) @comment")
+        .expect("Failed to write query");
+    fs::set_permissions(&query_dir, fs::Permissions::from_mode(0o000))
+        .expect("Failed to make query dir unreadable");
+    if fs::read_dir(&query_dir).is_ok() {
+        fs::set_permissions(&query_dir, fs::Permissions::from_mode(0o700))
+            .expect("Failed to restore query dir permissions");
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply tests/test_cli.rs]: Fixed in 954769afc: permission-based tests now consume every read_dir entry before deciding the directory is enumerable.

File: tests/test_cli.rs [hunk 12]
```diff
@@ -1209,6 +1209,820 @@ fn test_language_uninstall_all() {
...
+    fs::write(&parser, "fake").expect("Failed to write parser");
+    fs::write(nested.join("query.scm"), "(comment) @comment")
+        .expect("Failed to write nested query");
+    fs::set_permissions(&nested, fs::Permissions::from_mode(0o000))
+        .expect("Failed to make nested query unreadable");
+    if fs::read_dir(&nested).is_ok() {
+        fs::set_permissions(&nested, fs::Permissions::from_mode(0o700))
+            .expect("Failed to restore nested query permissions");
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply tests/test_cli.rs]: Fixed in 954769afc: permission-based tests now consume every read_dir entry before deciding the directory is enumerable.

File: tests/test_cli.rs [hunk 13]
```diff
@@ -1209,6 +1209,820 @@ fn test_language_uninstall_all() {
...
+        "ok\n",
+    )
+    .expect("Failed to mark backup ownership");
+    fs::set_permissions(&nested, fs::Permissions::from_mode(0o000))
+        .expect("Failed to make backup subtree unreadable");
+    if fs::read_dir(&nested).is_ok() {
+        fs::set_permissions(&nested, fs::Permissions::from_mode(0o700))
+            .expect("Failed to restore backup permissions");
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply tests/test_cli.rs]: Fixed in 954769afc: permission-based tests now consume every read_dir entry before deciding the directory is enumerable.

File: tests/test_cli.rs [hunk 14]
```diff
@@ -1209,6 +1209,820 @@ fn test_language_uninstall_all() {
...
+    fs::create_dir_all(&cache_dir).expect("Failed to create unrelated cache dir");
+    let parser = parser_dir.join(format!("lua.{}", std::env::consts::DLL_EXTENSION));
+    fs::write(&parser, "fake").expect("Failed to write parser");
+    fs::set_permissions(&cache_dir, fs::Permissions::from_mode(0o000))
+        .expect("Failed to make unrelated cache unreadable");
+    if fs::read_dir(&cache_dir).is_ok() {
+        fs::set_permissions(&cache_dir, fs::Permissions::from_mode(0o700))
+            .expect("Failed to restore cache permissions");
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply tests/test_cli.rs]: Fixed in 954769afc: permission-based tests now consume every read_dir entry before deciding the directory is enumerable.

**Comment by atusy** [comment]
@copilot review

File: src/bin/main.rs [hunk 15]
```diff
@@ -480,13 +484,214 @@ fn installed_query_language_name(path: &Path) -> Option<String> {
...
+            let file_type = entry
+                .file_type()
+                .map_err(|e| format!("cannot inspect query entry '{}': {e}", path.display()))?;
+            // remove_dir_all does not follow symlinks, so only real child
+            // directories need traversal preflight.
+            if file_type.is_dir() {
+                pending.push(path);
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/bin/main.rs:528]: Fixed in 675162ab9. Query-tree preflight now treats NotFound from directory open, entry iteration, and file-type inspection as an idempotent concurrent disappearance while preserving all other I/O failures.

File: src/install/queries.rs [hunk 16]
```diff
@@ -791,6 +842,29 @@ fn backup_is_owned(path: &Path) -> bool {
...
+        return Ok(false);
+    }
+    let Some(highlights) = regular_file_or_absent(&path.join("highlights.scm"))? else {
+        return Ok(false);
+    };
+    if !highlights.is_file() {
+        return Ok(false);
+    }
+    let marker = regular_file_or_absent(&path.join(QUERY_INSTALL_COMPLETE_MARKER))?;
+    Ok(marker.is_some_and(|metadata| metadata.is_file()) || highlights.len() > 0)
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/queries.rs:879]: Fixed in 675162ab9. Backup ownership/completeness markers now use symlink_metadata, so symlinks are not accepted as regular marker files.

**Comment by atusy** [comment]
@copilot review

**Comment by atusy** [comment]
@copilot review

**Comment by atusy** [comment]
@copilot review

**Comment by atusy** [comment]
@copilot review

File: src/install/queries.rs [hunk 17]
```diff
@@ -487,9 +487,13 @@ pub fn recover_interrupted_query_installs(queries_parent: &Path) -> Result<(), Q
...
-    for entry in entries.flatten() {
+    for entry in entries {
+        let entry = entry?;
         let path = entry.path();
-        if !path.is_dir() {
+        // Recovery artifacts must be real directories. Do not follow a
+        // staging/backup symlink outside the queries root, and propagate
+        // entry-type errors instead of silently treating them as absent.
+        if !entry.file_type()?.is_dir() {
             continue;
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/queries.rs:506]: Fixed in 4d99c340f. Recovery scans now skip NotFound from entry iteration, file type, backup completeness, and modified-time inspection while preserving all other I/O failures.

**Comment by atusy** [comment]
@copilot review

**Comment by atusy** [comment]
Closing due to scope bloat: 33 commits / +1401 lines accumulated from iterative self-review findings, covering many low-probability edge cases (malformed recovery field digits, backup-metadata ownership checks, etc.). The core fix (missing-vs-unreadable directory distinction for `uninstall` scans) is still needed — issue #767 stays open. Recommend re-approaching as a smaller, focused PR plus separate targeted follow-ups for the hardening edge cases.
